File: app/src/main/java/com/blog/demo/application/PopupWindowActivity.java
package com.blog.demo.application;

import android.app.Activity;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.CheckBox;
import android.widget.PopupWindow;

import com.blog.demo.LogTool;
import com.blog.demo.R;

public class PopupWindowActivity extends Activity implements View.OnClickListener {
    private static final String LOG_TAG = "PopupWindowActivity";
    private CheckBox mCbTouchable, mCbFocusable, mCbOutsideTouchable;
    private PopupWindow mPopupWindow;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_application_popup_window);

        findViewById(R.id.btn_show_at_location).setOnClickListener(this);
        mCbTouchable = findViewById(R.id.cb_touchable);
        mCbFocusable = findViewById(R.id.cb_focusable);
        mCbOutsideTouchable = findViewById(R.id.cb_outside_touchable);
        findViewById(R.id.btn_show_drop_down).setOnClickListener(this);
    }

    private PopupWindow createPopupWindow() {
        PopupWindow popupWindow = new PopupWindow(this);
        // 设置宽度
        popupWindow.setWidth(ViewGroup.LayoutParams.MATCH_PARENT);
        // 设置高度
        popupWindow.setHeight(ViewGroup.LayoutParams.WRAP_CONTENT);
        // 设置背景
        popupWindow.setBackgroundDrawable(new ColorDrawable(getResources().getColor(R.color.colorProduct)));
        View view = getLayoutInflater().inflate(R.layout.popup_window_picture, null, false);
        view.findViewById(R.id.tv_take_photo).setOnClickListener(this);
        view.findViewById(R.id.tv_take_picture).setOnClickListener(this);
        view.findViewById(R.id.tv_cancel).setOnClickListener(this);
        // 设置界面
        popupWindow.setContentView(view);
        // true时界面可点
        popupWindow.setTouchable(true);
        // true时PopupWindow处理返回键
        popupWindow.setFocusable(true);
        // true时点击外部消失，如果touchable为false时，点击界面也消失
        popupWindow.setOutsideTouchable(true);

        // dismiss监听器
        popupWindow.setOnDismissListener(new PopupWindow.OnDismissListener() {
            @Override
            public void onDismiss() {
                LogTool.logi(LOG_TAG, "onDismiss");
                backgroundAlpha(1);
            }
        });
        return popupWindow;
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.btn_show_at_location) {
            PopupWindow popupWindow = createPopupWindow();
            // true时界面可点
            popupWindow.setTouchable(mCbTouchable.isChecked());
            // true时PopupWindow处理返回键
            popupWindow.setFocusable(mCbFocusable.isChecked());
            // true时点击外部消失，如果touchable为false时，点击界面也消失
            popupWindow.setOutsideTouchable(mCbOutsideTouchable.isChecked());
            popupWindow.setAnimationStyle(R.style.popup_window_animation_style);
//            backgroundAlpha(0.5f);
            popupWindow.showAtLocation(v, Gravity.BOTTOM, 0, 0);

            mPopupWindow = popupWindow;
        } else if (v.getId() == R.id.btn_show_drop_down) {
            PopupWindow popupWindow = createPopupWindow();
            // true时界面可点
            popupWindow.setTouchable(mCbTouchable.isChecked());
            // true时PopupWindow处理返回键
            popupWindow.setFocusable(mCbFocusable.isChecked());
            // true时点击外部消失，如果touchable为false时，点击界面也消失
            popupWindow.setOutsideTouchable(mCbOutsideTouchable.isChecked());
            popupWindow.showAsDropDown(v);

            mPopupWindow = popupWindow;
        } else if ((v.getId() == R.id.tv_take_photo)
                || (v.getId() == R.id.tv_take_picture)
                || (v.getId() == R.id.tv_cancel)) {
            LogTool.logi(LOG_TAG, "PopupWindow onClick");
            mPopupWindow.dismiss();
        }
    }

    private void backgroundAlpha(float bgAlpha) {
        WindowManager.LayoutParams lp = getWindow().getAttributes();
        lp.alpha = bgAlpha; //0.0-1.0
        getWindow().setAttributes(lp);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);
    }

}

File: app/src/main/java/com/blog/demo/feature/popup/PopupWindowActivity.kt
package com.blog.demo.feature.popup

import android.app.Activity
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.*
import com.blog.demo.LogTool.logi
import com.blog.demo.R

class PopupWindowActivity : Activity(), View.OnClickListener {

    private val LOG_TAG = "PopupWindowActivity"

    private lateinit var mCbTouchable: CheckBox
    private lateinit var mCbFocusable: CheckBox
    private lateinit var mCbOutsideTouchable: CheckBox

    private var mPopupWindow: PopupWindow? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_feature_popup_popup_window)

        findViewById<Button>(R.id.btn_show_at_location).setOnClickListener(this)

        mCbTouchable = findViewById(R.id.cb_touchable)
        mCbFocusable = findViewById(R.id.cb_focusable)
        mCbOutsideTouchable = findViewById(R.id.cb_outside_touchable)

        findViewById<Button>(R.id.btn_show_drop_down).setOnClickListener(this)
    }

    private fun createPopupWindow(): PopupWindow {
        val popupWindow = PopupWindow(this)
        // 设置宽度
        popupWindow.width = ViewGroup.LayoutParams.MATCH_PARENT
        // 设置高度
        popupWindow.height = ViewGroup.LayoutParams.WRAP_CONTENT
        // 设置背景
        popupWindow.setBackgroundDrawable(ColorDrawable(resources.getColor(R.color.colorProduct)))

        val view: View = layoutInflater.inflate(R.layout.popup_window_picture, null, false)
        view.findViewById<TextView>(R.id.tv_take_photo).setOnClickListener(this)
        view.findViewById<TextView>(R.id.tv_take_picture).setOnClickListener(this)
        view.findViewById<TextView>(R.id.tv_cancel).setOnClickListener(this)

        // 设置界面
        popupWindow.contentView = view
        // true时界面可点
        popupWindow.isTouchable = true
        // true时PopupWindow处理返回键
        popupWindow.isFocusable = true
        // true时点击外部消失，如果touchable为false时，点击界面也消失
        popupWindow.isOutsideTouchable = true

        // dismiss监听器
        popupWindow.setOnDismissListener {
            logi(LOG_TAG, "onDismiss")
            backgroundAlpha(1f)
        }
        return popupWindow
    }

    override fun onClick(v: View) {
        if (v.id == R.id.btn_show_at_location) {
            mPopupWindow = createPopupWindow()
            // true时界面可点
            mPopupWindow?.isTouchable = mCbTouchable.isChecked
            // true时PopupWindow处理返回键
            mPopupWindow?.isFocusable = mCbFocusable.isChecked
            // true时点击外部消失，如果touchable为false时，点击界面也消失
            mPopupWindow?.isOutsideTouchable = mCbOutsideTouchable.isChecked
            mPopupWindow?.animationStyle = R.style.popup_window_animation_style
            //            backgroundAlpha(0.5f);
            mPopupWindow?.showAtLocation(v, Gravity.BOTTOM, 0, 0)
        } else if (v.id == R.id.btn_show_drop_down) {
            mPopupWindow = createPopupWindow()
            // true时界面可点
            mPopupWindow?.isTouchable = mCbTouchable.isChecked
            // true时PopupWindow处理返回键
            mPopupWindow?.isFocusable = mCbFocusable.isChecked
            // true时点击外部消失，如果touchable为false时，点击界面也消失
            mPopupWindow?.isOutsideTouchable = mCbOutsideTouchable.isChecked
            mPopupWindow?.showAsDropDown(v)
        } else if (v.id == R.id.tv_take_photo) {
            Toast.makeText(this, "take photo", Toast.LENGTH_LONG).show()
        } else if (v.id == R.id.tv_take_picture) {
            Toast.makeText(this, "take picture", Toast.LENGTH_LONG).show()
        } else if (v.id == R.id.tv_cancel) {
            Toast.makeText(this, "cancel", Toast.LENGTH_LONG).show()
            mPopupWindow?.dismiss()
        }
    }

    private fun backgroundAlpha(bgAlpha: Float) {
        val lp = window.attributes
        lp.alpha = bgAlpha //0.0-1.0
        window.attributes = lp
        window.addFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
    }

}